Run CNI plugins in the host mount namespace

ivan4th opened the pull request:
As of now, a hack is being used with nsenter-based wrapper scripts being generated for every file in `/opt/cni/bin` which may not be compatible with all CNI plugins. The script makes sure the plugin runs in the host network namespace (it's invoked from TapFDSource so it may start in another netns). It also depends on CNI plugins not requiring anything from the host filesystem besides `/etc/cni` and `/opt/cni/bin`, which may fail in some cases.
Let's just handle the namespaces in C/Go code instead.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/mirantis/virtlet/561)
<!-- Reviewable:end -->

jellonek commented:
Reviewed 13 of 13 files at r1.
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[cmd/vmwrapper/vmwrapper.go, line 73 at r1](https://reviewable.io:443/reviews/mirantis/virtlet/561#-L4KmY3RLa-RPpgCZHU_:-L4KmY3Sq8HLWL8EckNN:b8t5uix) ([raw file](https://github.com/mirantis/virtlet/blob/0b74121273ae0e658abf5e4fa08ceef2069a66e6/cmd/vmwrapper/vmwrapper.go#L73)):*
> ```Go
> 
> 	// configure glog (apparently no better way to do it ...)
> 	flag.CommandLine.Parse([]string{"-v=3", "-alsologtostderr=true"})
> ```

Btw. looks like we are also logging to file there - will change that in followup.

---

*[pkg/cni/client.go, line 157 at r1](https://reviewable.io:443/reviews/mirantis/virtlet/561#-L4KsctTVJdsJF4V4Pfn:-L4KsctTVJdsJF4V4Pfo:bytlgki) ([raw file](https://github.com/mirantis/virtlet/blob/0b74121273ae0e658abf5e4fa08ceef2069a66e6/pkg/cni/client.go#L157)):*
> ```Go
> 		"K8S_ANNOT", `{"cni": "calico"}`,
> 	})
> 	glog.V(3).Infof("AddSandboxToNetwork: req.PodId %q, req.PodName %q, req.PodNs %q, runtime config:\n%s",
> ```

Looks like too big mass rename. Prefixes in this message are useless.

---

*[pkg/cni/client.go, line 161 at r1](https://reviewable.io:443/reviews/mirantis/virtlet/561#-L4KsitIoDrErxR18JrC:-L4KsitIoDrErxR18JrD:b2nreh6) ([raw file](https://github.com/mirantis/virtlet/blob/0b74121273ae0e658abf5e4fa08ceef2069a66e6/pkg/cni/client.go#L161)):*
> ```Go
> 	result, err := c.cniConfig.AddNetworkList(c.netConfigList, rtConf)
> 	if err == nil {
> 		glog.V(3).Infof("AddSandboxToNetwork: req.PodId %q, req.PodName %q, req.PodNs %q: result:\n%s",
> ```

Same as above.

---

*[pkg/cni/client.go, line 164 at r1](https://reviewable.io:443/reviews/mirantis/virtlet/561#-L4Ksk_hDN7pWlbswywj:-L4Ksk_hDN7pWlbswywk:b2nreh6) ([raw file](https://github.com/mirantis/virtlet/blob/0b74121273ae0e658abf5e4fa08ceef2069a66e6/pkg/cni/client.go#L164)):*
> ```Go
> 			req.PodId, req.PodName, req.PodNs, spew.Sdump(result))
> 	} else {
> 		glog.V(3).Infof("AddSandboxToNetwork: req.PodId %q, req.PodName %q, req.PodNs %q: error: %v",
> ```

Same as above.

---

*[pkg/cni/client.go, line 176 at r1](https://reviewable.io:443/reviews/mirantis/virtlet/561#-L4KtD2F51R07U902g47:-L4KtD2F51R07U902g48:b7zxr1d) ([raw file](https://github.com/mirantis/virtlet/blob/0b74121273ae0e658abf5e4fa08ceef2069a66e6/pkg/cni/client.go#L176)):*
> ```Go
> 
> func handleRemoveSandboxFromNetwork(arg interface{}) (interface{}, error) {
> 	req := arg.(*cniRequest)
> ```

This is sooo hacky ;)

---

*[pkg/cni/client.go, line 182 at r1](https://reviewable.io:443/reviews/mirantis/virtlet/561#-L4Ksn233w1-pLW-FRvP:-L4Ksn233w1-pLW-FRvQ:bqo5gdv) ([raw file](https://github.com/mirantis/virtlet/blob/0b74121273ae0e658abf5e4fa08ceef2069a66e6/pkg/cni/client.go#L182)):*
> ```Go
> 	}
> 
> 	glog.V(3).Infof("RemoveSandboxFromNetwork: req.PodId %q, req.PodName %q, req.PodNs %q", req.PodId, req.PodName, req.PodNs)
> ```

Again.

---

*[pkg/cni/client.go, line 185 at r1](https://reviewable.io:443/reviews/mirantis/virtlet/561#-L4Ksoiqndhv4k-0QqOu:-L4Ksoiqndhv4k-0QqOv:bqo5gdv) ([raw file](https://github.com/mirantis/virtlet/blob/0b74121273ae0e658abf5e4fa08ceef2069a66e6/pkg/cni/client.go#L185)):*
> ```Go
> 	err = c.cniConfig.DelNetworkList(c.netConfigList, c.cniRuntimeConf(req.PodId, req.PodName, req.PodNs))
> 	if err == nil {
> 		glog.V(3).Infof("RemoveSandboxFromNetwork: req.PodId %q, req.PodName %q, req.PodNs %q: success",
> ```

Again.

---

*[pkg/cni/client.go, line 188 at r1](https://reviewable.io:443/reviews/mirantis/virtlet/561#-L4KsqDDVXmViaN_-hBi:-L4KsqDDVXmViaN_-hBj:b8w8aql) ([raw file](https://github.com/mirantis/virtlet/blob/0b74121273ae0e658abf5e4fa08ceef2069a66e6/pkg/cni/client.go#L188)):*
> ```Go
> 			req.PodId, req.PodName, req.PodNs)
> 	} else {
> 		glog.V(3).Infof("RemoveSandboxFromNetwork: req.PodId %q, req.PodName %q, req.PodNs %q: error: %v",
> ```


And again :)

---

*[pkg/utils/nsfix.go, line 33 at r1](https://reviewable.io:443/reviews/mirantis/virtlet/561#-L4KtMHdquurZc4ncCI4:-L4KtMHdquurZc4ncCI5:bcnghd0) ([raw file](https://github.com/mirantis/virtlet/blob/0b74121273ae0e658abf5e4fa08ceef2069a66e6/pkg/utils/nsfix.go#L33)):*
> ```Go
> )
> 
> type NsFixReexecHandler func(arg interface{}) (interface{}, error)
> ```

Missing docstring.

---

*[pkg/utils/nsfix.go, line 73 at r1](https://reviewable.io:443/reviews/mirantis/virtlet/561#-L4KtSmgZbvklhdseWz8:-L4KtSmgZbvklhdseWz9:bkr5f6a) ([raw file](https://github.com/mirantis/virtlet/blob/0b74121273ae0e658abf5e4fa08ceef2069a66e6/pkg/utils/nsfix.go#L73)):*
> ```Go
> 	}
> 	// configure glog (apparently no better way to do it ...)
> 	flag.CommandLine.Parse([]string{"-v=" + logLevelStr, "-alsologtostderr=true"})
> ```

Logging also to file. Can be removed in follow up.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/virtlet/561)*
<!-- Sent from Reviewable.io -->

ivan4th commented:
Review status: 9 of 13 files reviewed at latest revision, 10 unresolved discussions.

---

*[cmd/vmwrapper/vmwrapper.go, line 73 at r1](https://beta.reviewable.io/reviews/mirantis/virtlet/561#-L4KmY3RLa-RPpgCZHU_:-L4jYdzdDJJJHks2LkTX:b-xb2mys) ([raw file](https://github.com/mirantis/virtlet/blob/0b74121273ae0e658abf5e4fa08ceef2069a66e6/cmd/vmwrapper/vmwrapper.go#L73)):*
<details><summary><i>Previously, jellonek (Piotr Skamruk) wrote…</i></summary><blockquote>

Btw. looks like we are also logging to file there - will change that in followup.
</blockquote></details>

Fixed it now.

---

*[pkg/cni/client.go, line 157 at r1](https://beta.reviewable.io/reviews/mirantis/virtlet/561#-L4KsctTVJdsJF4V4Pfn:-L4jYiND9X5R1LnYNf_P:b-896fix) ([raw file](https://github.com/mirantis/virtlet/blob/0b74121273ae0e658abf5e4fa08ceef2069a66e6/pkg/cni/client.go#L157)):*
<details><summary><i>Previously, jellonek (Piotr Skamruk) wrote…</i></summary><blockquote>

Looks like too big mass rename. Prefixes in this message are useless.
</blockquote></details>

Done.

---

*[pkg/cni/client.go, line 161 at r1](https://beta.reviewable.io/reviews/mirantis/virtlet/561#-L4KsitIoDrErxR18JrC:-L4jYivH-oE-qgJd5T8C:b-896fix) ([raw file](https://github.com/mirantis/virtlet/blob/0b74121273ae0e658abf5e4fa08ceef2069a66e6/pkg/cni/client.go#L161)):*
<details><summary><i>Previously, jellonek (Piotr Skamruk) wrote…</i></summary><blockquote>

Same as above.
</blockquote></details>

Done.

---

*[pkg/cni/client.go, line 164 at r1](https://beta.reviewable.io/reviews/mirantis/virtlet/561#-L4Ksk_hDN7pWlbswywj:-L4jYjPQ68oEm0fMZki2:b-896fix) ([raw file](https://github.com/mirantis/virtlet/blob/0b74121273ae0e658abf5e4fa08ceef2069a66e6/pkg/cni/client.go#L164)):*
<details><summary><i>Previously, jellonek (Piotr Skamruk) wrote…</i></summary><blockquote>

Same as above.
</blockquote></details>

Done.

---

*[pkg/cni/client.go, line 176 at r1](https://beta.reviewable.io/reviews/mirantis/virtlet/561#-L4KtD2F51R07U902g47:-L4jYkBO0sHrpiS2cGxS:b-mcsq8l) ([raw file](https://github.com/mirantis/virtlet/blob/0b74121273ae0e658abf5e4fa08ceef2069a66e6/pkg/cni/client.go#L176)):*
<details><summary><i>Previously, jellonek (Piotr Skamruk) wrote…</i></summary><blockquote>

This is sooo hacky ;)
</blockquote></details>

Well not quite. When you register the reexec handler it's guaranteed to get specific arg type, and I didn't bother with using `reflect` because nsfix is used just a few times in the code. Added this to NsFixReexecHandler docs.

---

*[pkg/cni/client.go, line 182 at r1](https://beta.reviewable.io/reviews/mirantis/virtlet/561#-L4Ksn233w1-pLW-FRvP:-L4jYyHp2D8mpLquN4Ow:b-896fix) ([raw file](https://github.com/mirantis/virtlet/blob/0b74121273ae0e658abf5e4fa08ceef2069a66e6/pkg/cni/client.go#L182)):*
<details><summary><i>Previously, jellonek (Piotr Skamruk) wrote…</i></summary><blockquote>

Again.
</blockquote></details>

Done.

---

*[pkg/cni/client.go, line 185 at r1](https://beta.reviewable.io/reviews/mirantis/virtlet/561#-L4Ksoiqndhv4k-0QqOu:-L4jYyt14kAvXxgLDUsf:b-896fix) ([raw file](https://github.com/mirantis/virtlet/blob/0b74121273ae0e658abf5e4fa08ceef2069a66e6/pkg/cni/client.go#L185)):*
<details><summary><i>Previously, jellonek (Piotr Skamruk) wrote…</i></summary><blockquote>

Again.
</blockquote></details>

Done.

---

*[pkg/cni/client.go, line 188 at r1](https://beta.reviewable.io/reviews/mirantis/virtlet/561#-L4KsqDDVXmViaN_-hBi:-L4jYzRtD_f8B4H7OHh2:b-896fix) ([raw file](https://github.com/mirantis/virtlet/blob/0b74121273ae0e658abf5e4fa08ceef2069a66e6/pkg/cni/client.go#L188)):*
<details><summary><i>Previously, jellonek (Piotr Skamruk) wrote…</i></summary><blockquote>

And again :)
</blockquote></details>

Done.

---

*[pkg/utils/nsfix.go, line 33 at r1](https://beta.reviewable.io/reviews/mirantis/virtlet/561#-L4KtMHdquurZc4ncCI4:-L4jZHQLCHdGuekXrg6N:b-896fix) ([raw file](https://github.com/mirantis/virtlet/blob/0b74121273ae0e658abf5e4fa08ceef2069a66e6/pkg/utils/nsfix.go#L33)):*
<details><summary><i>Previously, jellonek (Piotr Skamruk) wrote…</i></summary><blockquote>

Missing docstring.
</blockquote></details>

Done.

---

*[pkg/utils/nsfix.go, line 73 at r1](https://beta.reviewable.io/reviews/mirantis/virtlet/561#-L4KtSmgZbvklhdseWz8:-L4jZIEz9lmVz7HjQz3t:b-xb2mys) ([raw file](https://github.com/mirantis/virtlet/blob/0b74121273ae0e658abf5e4fa08ceef2069a66e6/pkg/utils/nsfix.go#L73)):*
<details><summary><i>Previously, jellonek (Piotr Skamruk) wrote…</i></summary><blockquote>

Logging also to file. Can be removed in follow up.
</blockquote></details>

Fixed it now.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/mirantis/virtlet/561)*
<!-- Sent from Reviewable.io -->

jellonek commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 4 of 4 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/mirantis/virtlet/561#-:-L4jZyCm6p_Hs-eGZmr1:bnfp4nl)*
<!-- Sent from Reviewable.io -->

lukaszo commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[deploy/virtlet-ds.yaml, line 127 at r2](https://reviewable.io:443/reviews/mirantis/virtlet/561#-L4kP7J7v42o3Nu766bz:-L4kP7J7v42o3Nu766c-:b-c87vcq) ([raw file](https://github.com/mirantis/virtlet/blob/8541b6efb3272473a6000ae381fabaf43ca13b6f/deploy/virtlet-ds.yaml#L127)):*
> ```YAML
>           # and CNI plugins (which run in the host netns) can work
>           # with network namespaces mounted there
>         - mountPath: /var/run/netns:shared
> ```

do we need this? /var/run is a symlink to /var(at least on kdc) which already is shared

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/virtlet/561)*
<!-- Sent from Reviewable.io -->

lukaszo commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/virtlet/561#-:-L4kTapVw7BH0pmrhAKP:bnfp4nl)*
<!-- Sent from Reviewable.io -->
